### PR TITLE
[Feature] Support for PMD Copy/Paste Detection

### PR DESCRIPTION
**Pull request by kloessst**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

Clones are a threat to the maintainability of software systems. Pieces of code which are used often should be encapsulated and made available to other components by providing well documented interfaces (Don't repeat yourself, Single responsibility principle). With this PR, code clones can be automatically detected and punished by configuring the grading for programming exercises.

### Description

- Add `PMD CPD` tool, which automatically extends the build plan if SCA is enabled
- Add Copy Paste detection category
- Add default configuration for PMD CPD. Only duplication with a minimum size of 70 tokens are reported.

### Steps for Testing

**Can only be tested on TS2 at the moment, as this PR relies on a new version of the Notification Plugin, which is already deployed on Jenkins**
1. Create a programming exercise with SCA enabled
2. Participate
3. Submit code with contains duplications (must be at least 70 tokens long, this can be changed in the pom, the BubbleSort algorithm should suffice)
--> Check that the code duplication is detected

### Test Coverage

Nothing changed

### Screenshots

image

### CODE REVIEW

**Review by fde312 — COMMENTED** [review]
I added the same method for all 3 classes and got the following issues:

image

I would expect just to get the 3 issues with the 3 files and not additionally 2 issues with `BubbleSort` and `Client. Is this intended @kloessst ?

**Comment by kloessst** [comment]
> I added the same method for all 3 classes and got the following issues:
> 
> image
> 
> I would expect just to get the 3 issues with the 3 files and not additionally 2 issues with `BubbleSort` and `Client. Is this intended @kloessst ?

I guess it finds one more duplicate token/line as in the other duplication and creates another issue for this. As the copy paste detection is a blackbox for us, we can't change this. 
The reporting of multiple feedback elements for one unique duplication is wanted and approved by @krusche. This allows us to highlight the locations in the code editor and to punish multiple clones of the same code harder.

I will set this PR to the draft state, until it can be tested on both TS1 and TS2.

**Review by MaisiKoleni — DISMISSED** [review]
Code looks good to me. Tested on TS 2, works fine.

File: src/main/resources/templates/java/test/projectTemplate/pom.xml [hunk 1]
```diff
@@ -100,6 +100,15 @@
                         <version>6.28.0</version>
                     </dependency>
                 </dependencies>
+                <configuration>
```
Comment (fde312) [inline src/main/resources/templates/java/test/projectTemplate/pom.xml:103]: I think we now have two `pom.xml` files right? One for eclipse and one for maven. This change should also be done for the other right?

**Review by fde312 — CHANGES_REQUESTED** [review]

**Review by fde312 — APPROVED** [review]
CPD works for me: 
image

**Comment by MaisiKoleni** [comment]
Is there any reason for the changes in c7b25bc?

**Comment by kloessst** [comment]
> 
> 
> Is there any reason for the changes in [c7b25bc](https://github.com/ls1intum/Artemis/commit/c7b25bc50c1f4cce63be52db8cb5740d34732b40)?

I did some testing, playing around with the options. If you activate `ignoreIdentifiers`, a lot of tokens e.g. variable names, method names, won't be counted any longer. In turn, the `minimumTokens` must be reduced significantly, otherwise no duplications will be found. Lowering the `minimumTokens` too much, increases the risk of false positives.

I tested this with the MergeSort implementation of the template and using `ignoreIdentifiers` with 70 `minimumTokens`, copying the whole MergeSort implementation won't be detected. Reducing `minimumTokens` to 30 detects a if block duplication within the implementation which is a false positive. For 50 `minimumTokens` nothing is detected.

I try to find a good default configuration but in the end, each instructor is able to play around with these settings as well. Let me know if you found a better solution. Then we can use this as the default.

**Review by MaisiKoleni — APPROVED** [review]
The CPD still works fine, also tested with Maven Java projects. It is a bit sad to lose the "ignore identifiers" option which makes it a bit easier to circumvent but it is a limitation of PMD and there are other ways to hide C&P. (This makes grading it more problematic because avoid the CPD messages can lead to worse code, especially is students are inexperienced). But I cannot think of a better way.

**Review by krusche — APPROVED** [review]

**Pull request merged**